### DISCUSSION TITLE
Can i get some context about this issue?

### DISCUSSION THREAD

**Discussion by octonawish-akcodes** [kcl-lang/kpm#261] [General]
[This issue](https://github.com/kcl-lang/kpm/issues/246) is talking about this:

> Describe the feature you'd like:
I would like to propose a version management feature for our package manager. This feature should allow users to easily view, install, upgrade, and downgrade different versions of packages. It should also handle dependencies version conflicts between different versions of the same package to ensure that all dependencies are met when installing, upgrading, or downgrading a package.

Here for `view` scenario what are we considering from kpm command something like this --> `kpm show mod` which will show the current installed modules?

For installing, upgrading, downgrading a specific module we already install the modules dependencies as `kpm add <somemodule@<followed by version>>` which particularly gives the user to install whatever version he wants right, Also i tried running this command to check the installing, upgrading(by providing the specific version), downgrading(by providing the version) the kpm add overwrites the dependencies in the `kcl.mod` file, This kinda creates a confusion to me with what @zong-zhe is suggesting here ` This feature should allow users to easily view, install, upgrade, and downgrade different versions of packages.` 

Also when i try to recreate this `It should also handle dependencies version conflicts between different versions of the same package` it just overwrites the package version, So I dont quite understand this scenario here as well. 

@Peefy @zong-zhe

**Comment by octonawish-akcodes** [comment]
any suggestions about how I can create conflicts of version? @zong-zhe

**Comment by zong-zhe** [comment]
Hi @octonawish-akcodes 🙂

I apologize for the late response, I’m in the middle of something right now.
Let me provide some more details, there seems to be a misunderstanding here.😆

> Also when i try to recreate this `It should also handle dependencies version conflicts between different versions of the same package` it just overwrites the package version, So I dont quite understand this scenario here as well.

The kpm sometimes overwrites versions, which is caused by the lack of package version management. Because kpm currently does not manage package versions, it just recursively downloads all dependencies, therefore, dependencies added later will overwrite those added earlier who have the same version.

If version management is completed, kpm, in the process of adding dependencies, if it needs to add different versions of the same package, then kpm will calculate a compatible version through version management and download it. If a compatible version cannot be found, then an error should be thrown, notifying the user that your project depends on two different and incompatible versions of the same package.

**Comment by zong-zhe** [comment]
@octonawish-akcodes Additionally, I just created a PreTest https://github.com/kcl-lang/kpm/issues/263, adding a feature to the `kcl mod` command similar to `go mod graph`, which outputs the dependency graph of the current KCL package. This feature will be of great help in version management.

**Comment by zong-zhe** [comment]
Hi @octonawish-akcodes,

More details added here - https://github.com/kcl-lang/kpm/issues/246
